feat: Adds a key-value store endpoint for Superset

michael-s-molina opened the pull request:
### SUMMARY
This PR adds a key-value store endpoint for Superset with secure key generation and management of store entries. 

Keys are generated on the server-side using the https://docs.python.org/3/library/secrets.html#secrets.token_urlsafe function and the value can be any JSON supported text.

Each value can have an associated duration.  A Celery cleanup task is responsible for managing expired entries in the key-value store. 

The user can also set a configuration to reset the duration upon retrieval to preserve actively accessed entries. 

This endpoint can be used to solve a bunch of uses cases for Superset. Each client of this endpoint should carefully consider security, normalization, and ownership implications when deciding to store some type of information into the key-value store. Some possible use cases for this endpoint are listed below:
- Profiles, preferences, and configurations management
- Large objects storage (URL parameters, application state, drafts, etc)
- Caching

We currently have an endpoint at `/kv/store`, but it has serial/monotonic ids which allow a user to iterate through other people’s saved states. That endpoint is also using the deprecated API system on `superset/views`, so it would need to be refactored anyway. Another important point is that the `/kv/store` endpoint does not contain any management capabilities regarding entries duration, ownership, and creation timestamps, which makes automatic sanitization harder.

The security model of the entry point is inspired by the Google Drive sharing model and it’s intended to support similar use cases. By default, any added value to the key-value store is accessible only to the user who created the value. This can be altered using the `anyone_with_key`, `editor_*`,  and `viewer_*` configurations. The `anyone_with_key` option accepts the `editor` and `viewer` values indicating that anyone with the key can edit or view the store value. This is useful when storing non-sensitive information that needs to be shared among users. The `editor_*` and `viewer_*`  configurations allow specific access restrictions to the key-value entries like granting read or write permissions to specific users or roles. In the case of access configuration conflicts, the more restrictive rule should always apply. An example would be `anyone_with_key: ‘editor’` and `editor_users: [1, 2, 3]`. In this case, only users 1, 2, and 3 would have editor access. 

Here's a pseudo-code of the endpoint operations and their configurations. For the complete status codes and possible responses check the Swagger documentation:

```
POST `api/v1/key_value_store`

BODY 
{
   // The duration of the value in the key store.
   // If no duration is specified the value won't expire.
   duration_ms?: <number>,

   // If the duration should be reset when the value is retrieved. 
   // This is useful if you wish to expire unused values but keep the ones that are actively retrieved.
   // True by default
   reset_duration_on_retrieval?: <boolean>,

   // Any type of JSON supported text.
   value: <string>,

   // Anyone with the key can edit or view the store value
   anyone_with_key?: <'editor', 'viewer'>

   // IDs of the users with editor access
   editor_users?: <[number]>

   // IDs of the roles with editor access
   editor_roles?: <[number]>

   // IDs of the users with viewer access
   viewer_users?: <[number]>

   // IDs of the roles with viewer access
   viewer_roles?: <[number]>
}

RESPONSE
{
   // The key generated on the server-side
   key: <string>
}
```

```
PUT `api/v1/key_value_store/<key>`

BODY - Same as POST body

RESPONSE
{
   // The result of the operation
   message: <string>
}
```

```
GET `api/v1/key_value_store/<key>`

RESPONSE
{
   // Any type of JSON supported text.
   value: <string>,
}
```

```
DELETE `api/v1/key_value_store/<key>`

RESPONSE
{
   // The result of the operation
   message: <string>
}
```

### TESTING INSTRUCTIONS
1 -  Execute Python tests
2 - All tests should pass

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

michael-s-molina committed:
Initial code

michael-s-molina committed:
Adds more props to the model

michael-s-molina committed:
Removes unused imports of core.py

zhaoyongjie commented on superset/key_value/api.py:
```diff
@@ -0,0 +1,154 @@
+    class_permission_name = "KeyValue"
+    resource_name = "key_value_store"
+    method_permission_name = MODEL_API_RW_METHOD_PERMISSION_MAP
+
```
include_route_methods = {RouteMethod.POST, RouteMethod.GET}

zhaoyongjie replied:
use this property and add uncomment `@protect()`
```
    allow_browser_login = True
```

michael-s-molina committed:
Adds the cleanup celery job and missing endpoints

michael-s-molina committed:
Generates the key using a SHA3_256 hash

michael-s-molina committed:
Ignores expired values on find_by_id

michael-s-molina committed:
Fixes the PUT endpoint

michael-s-molina committed:
Adds the Swagger docs

michael-s-molina committed:
Adjusts the tests

betodealmeida commented on superset/config.py:
```diff
@@ -732,6 +732,10 @@ class CeleryConfig:  # pylint: disable=too-few-public-methods
         },
+        "key_value.cleanup": {
+            "task": "key_value.cleanup",
+            "schedule": timedelta(seconds=10),
```
How long does it take to run the jobs? 10 seconds looks a bit aggressive to me, maybe we could have a default of ~1 minute to be conservative on resources?

michael-s-molina replied:
I forgot to change that 🤪. I was just using 10 seconds for testing but the default should be 24 hours.

betodealmeida commented on superset/key_value/api.py:
```diff
@@ -0,0 +1,324 @@
+              $ref: '#/components/responses/500'
+        """
+        if not request.is_json:
+            return self.response_400(message="Request is not JSON")
```
Can we make the error responses comply with SIP-40? https://github.com/apache/superset/issues/9194

We should return a JSON payload with more detail instead of a string whenever an error occurs.

michael-s-molina replied:
Yes! Thanks for pointing that out!

betodealmeida commented on superset/key_value/api.py:
```diff
@@ -0,0 +1,324 @@
+                str(ex),
+                exc_info=True,
+            )
+            return self.response_422(message=str(ex))
```
Same here an in other methods below.

betodealmeida commented on superset/key_value/commands/cleanup.py:
```diff
@@ -0,0 +1,47 @@
+    def run(self) -> None:
+        logger.info("Key value store cleanup starting")
+        with session_scope(nullpool=True) as session:
+           for keyValue in session.query(KeyValue).all():
+                if is_expired(keyValue):
```
It would be much better for performance to implement the logic of `is_expired` in SQL instead of Python, specially because we already have a query on line 39.

michael-s-molina replied:
Yes. In fact, I'm currently doing this exact change. I'm just trying to learn how to do that with SQL Alchemy since this is my first Python PR :sweat_smile:

betodealmeida replied:
Oh, good job then! I could never tell this was your first Python PR!

And yeah, doing SQL with SQLAlchemy is not fun... :-/

dpgaspar replied:
There must be some way to restrict this full table scan `session.query(KeyValue).all()`. What if, when we create/update a key we timestamp an expiration date time then filter by it?

michael-s-molina replied:
I'm on it. A DELETE with a WHERE clause and a simple condition will be sufficient.

betodealmeida commented on tests/integration_tests/key_value/api_tests.py:
```diff
@@ -0,0 +1,79 @@
+    def test_post(self):
+        key = self.post()
+        result = db.session.query(KeyValue).first()
+        self.assertEqual(key, str(result.key))
```
Nit: `pytest` works better with naked asserts, eg:

```suggestion
        assert key == str(result.key)
```

https://docs.pytest.org/en/latest/how-to/assert.html#assert

betodealmeida commented on tests/integration_tests/key_value/api_tests.py:
```diff
@@ -0,0 +1,79 @@
+        )
+        self.assertEqual(value, data.get("value"))
+
+    def test_get_retrieved_on(self):
```
Can you use `freezegun` and add a unit test checking that `retrived_on` is set to an expected value?

https://github.com/apache/superset/blob/4dc859f89e5668ed3f94cd1ef0532a301a3ab85a/tests/integration_tests/reports/scheduler_tests.py#L45

betodealmeida reviewed (changes requested):
I left a few comments... the most important is to return SIP-40 compliant errors, specially since this is a new API.

willbarrett reviewed (changes requested):
I'm concerned about some of the proposed usages - particularly access tokens, refresh tokens, and public key storage. These items should all be held in an encrypted system, not a plain-text field. That's a big security no-no.

I would also recommend adding a user_id to the key value store and only allowing retrieval of items by the same user. Otherwise this system potentially allows any user to read any stored key, which for most of the uses you recommend would constitute a security hole.

The original key value store was originally deprecated due to similar security concerns. Keep in mind that obscurity of a long key is not the same thing as security - even if the keys are hard to guess, we should have security controls on the individual keys.

michael-s-molina commented:
Hi @willbarrett. Thanks so much for helping with the review!

> I'm concerned about some of the proposed usages - particularly access tokens, refresh tokens, and public key storage. These items should all be held in an encrypted system, not a plain-text field. That's a big security no-no.

You're right about this. There are more secure structures for this type of information. I removed them from possible use cases in the PR description.

> I would also recommend adding a user_id to the key value store and only allowing retrieval of items by the same user. Otherwise this system potentially allows any user to read any stored key, which for most of the uses you recommend would constitute a security hole.
> The original key value store was originally deprecated due to similar security concerns. Keep in mind that obscurity of a long key is not the same thing as security - even if the keys are hard to guess, we should have security controls on the individual keys.

The key-value table has a `created_by` field for that purpose. One of the main uses of this store is to share content between users like an URL state, a draft, etc. Currently, the access is being controlled by the possession of a secure key. It's the same type of access control as Google Docs when you select the "Anyone with the link" type of sharing. We could increment this and also offer the second type of control where we can select which users are allowed to access the content, similar to Google Docs restricted sharing. What do you think?

willbarrett commented:
Thanks for the response @michael-s-molina - I think as Superset moves into more organizations we should default to closed in all cases, so I would support the closed-sharing model. The open-sharing model is pretty dangerous - it becomes easy to create a link that's accessible to everyone in a company, which I believe wouldn't be a desired behavior by most larger organizations as a default behavior. In Superset, URL parameters, application state, and cache can contain highly sensitive information so I think we should shy away from the open-sharing model in all cases.

michael-s-molina commented:
> Thanks for the response @michael-s-molina - I think as Superset moves into more organizations we should default to closed in all cases, so I would support the closed-sharing model. The open-sharing model is pretty dangerous - it becomes easy to create a link that's accessible to everyone in a company, which I believe wouldn't be a desired behavior by most larger organizations as a default behavior. In Superset, URL parameters, application state, and cache can contain highly sensitive information so I think we should shy away from the open-sharing model in all cases.

@willbarrett These are good points. I agree that changing the default to the restricted model is more appropriate. I also think we should support the "Anyone with the key" model because we have some resources like public dashboards where we can benefit from it and we don't need the whole security configuration part from the user. I'll ping @dpgaspar to discuss this and increment the PR with these requirements.  Thank you so much again!

willbarrett commented:
Thanks @michael-s-molina - if we do implement the "anyone with a key" model, we should throw some restrictions or confirmation around it so it's very clear to the user that they're about to share very widely. Something to think about on the UI-side of the house.

michael-s-molina commented:
> Thanks @michael-s-molina - if we do implement the "anyone with a key" model, we should throw some restrictions or confirmation around it so it's very clear to the user that they're about to share very widely. Something to think about on the UI-side of the house.

Definitely! We should be really clear about the security implications. I was just checking the Google Drive interface and we can follow the same pattern.

michael-s-molina committed:
Limits key size in database

michael-s-molina committed:
Use naked asserts

dpgaspar commented on superset/key_value/api.py:
```diff
@@ -0,0 +1,324 @@
+
+class KeyValueRestApi(BaseSupersetModelRestApi):
+    datamodel = SQLAInterface(KeyValue)
+    schema = KeyValueSchema()
```
nit: let's use the `add_model_schema` and `edit_model_schema` attrs, they already exist and their goal is exactly this

michael-s-molina replied:
Done

dpgaspar commented on superset/key_value/api.py:
```diff
@@ -0,0 +1,324 @@
+            content:
+              application/json:
+                schema:
+                    type: object
```
it's shorter to just use a reference to the marshmallow schema like this: https://github.com/apache/superset/blob/master/superset/dashboards/api.py#L427

michael-s-molina replied:
Done

dpgaspar commented on superset/key_value/api.py:
```diff
@@ -0,0 +1,324 @@
+            201:
+              description: The value was stored successfully. It returns the key to retrieve the value.
+              content:
+                application/json:
```
Create a marshmallow schema for the response and reference it here, you can register new marshmallow schemas like this: https://github.com/apache/superset/blob/master/superset/dashboards/api.py#L202
and then reference them like: https://github.com/apache/superset/blob/master/superset/dashboards/api.py#L323

michael-s-molina replied:
Done

dpgaspar commented on superset/key_value/commands/create.py:
```diff
@@ -0,0 +1,61 @@
+    def run(self) -> Model:
+        try:
+            secretKey = current_app.config["SECRET_KEY"]
+            key = sha3_256(secretKey.encode() + str(uuid4()).encode()).hexdigest()
```
probably does not make any diference to the solution itself, but I would prefer to keep the `SECRET_KEY` out of this, think secret key rotation, something that some organisations are already implementing (no impact to the solution but). I think UUID4 over sha3_256 is good enough, UUID4 possible combinations 2^128. We can increase this by using sha3_512 for example

michael-s-molina replied:
I think this is a good point. Will replace it with a random seed.

michael-s-molina replied:
I ended up using https://docs.python.org/3/library/secrets.html#secrets.token_urlsafe for the key generation 😉

dpgaspar commented on superset/key_value/commands/get.py:
```diff
@@ -0,0 +1,50 @@
+            model = KeyValueDAO.find_by_id(self._key)
+            if not model:
+                return None
+            setattr(model, 'retrieved_on', datetime.utcnow())
```
curious, are you avoiding auditmixin?

michael-s-molina replied:
I just didn't know about it 😆. Now I'm using it.

dpgaspar commented on superset/key_value/utils.py:
```diff
@@ -0,0 +1,27 @@
+from sqlalchemy import Boolean
+
+
+def is_expired(keyValue: KeyValue) -> Boolean:
```
Highly optional, even debatable, but we could implement this on the model itself

dpgaspar commented on superset/models/key_value.py:
```diff
@@ -0,0 +1,34 @@
+
+    __tablename__ = "key_value"
+    key = Column(String(256), primary_key=True)
+    value = Column(Text, nullable=False)
```
if `value` is potentially sensitive, we should encrypt this field

dpgaspar reviewed (commented):
left some comments, we should add more tests also

dpgaspar commented:
> Thanks @michael-s-molina - if we do implement the "anyone with a key" model, we should throw some restrictions or confirmation around it so it's very clear to the user that they're about to share very widely. Something to think about on the UI-side of the house.

Would definitely fall to restrict access to the owner of the key. but the K/V store goal is not clear yet or it's just too broad. Session management and caching are sensitive, caching values could potentially defeat dataset ownership and RBAC permissions.

We can make the ownership restriction optional and on by default behind a config key. Or discuss this further on a secure sharing model

michael-s-molina commented:
> left some comments, we should add more tests also

Thanks for the review @dpgaspar! I'll address all your comments

michael-s-molina committed:
Changes the key generation to use the secrets package

michael-s-molina commented:
@dpgaspar @willbarrett @betodealmeida I updated the PR description with the results of our last interactions about security and suggested use cases. I also added examples of the endpoint operations and their configurations.

@john-bodley I would really like your review too. 

@dpgaspar even though the names in the API are editor and viewer, I’m still going to use the owner nomenclature in the database to match our existing schema 😉 

You'll notice that the code is not reflecting the current description. I'm looking for more reviews on the security model before implementing it. If you can, just give a thumbs-up in this comment so I know you're ok with it or add a comment otherwise.

I'm really excited about my first Python PR so thank you all for the reviews. It's good to be back to full-stack development 🙌🏼

michael-s-molina committed:
Extends from Audit model and uses schemas in docs

michael-s-molina committed:
Improves the cleanup job

michael-s-molina committed:
Fixes the response status code docs

michael-s-molina committed:
Fixes exceptions.py formatting

michael-s-molina committed:
Fixes __init__.py formatting

michael-s-molina committed:
Updates the down_revision

michael-s-molina committed:
Makes the API SIP-40 compatible

michael-s-molina committed:
Adjusts frozen time

geido commented:
Awesome work @michael-s-molina!

michael-s-molina commented:
We had a meeting about the proposed solution in this PR and decided to take a slightly different approach to resolve the use cases. Exposing a generic key-value interface had the potential to induce users to store any type of information in the store,  diminishing the benefits of the current endpoints where each type of resource is represented by a specific path. We could also introduce some conflicts with the current security model in terms of access grants. One example would be to store information in the key-value store that is related to a dashboard. The dashboard already has configured access grants and re-defining those in each key-value entry had the potential for security problems. We decided to offer the same type of functionality under each currently defined endpoint to preserve the more typed nature of our endpoints and to leverage the existing security model. I'm closing this PR to preserve all the discussions and conclusions for historical reasons and I'll open a new one with the new approach.

A big thanks to all the reviewers that are helping shape this new feature.
